## Final_Project: Prospects of Machine learning Approaches in predicting new variants or new mutations in hCoV-2 spike sequences

In [ ]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')
## multiple outputs
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import minmax_scale
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder# example of a one hot encoding
from numpy import asarray# example of a one hot encoding
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PowerTransformer
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm
from scipy.stats import boxcox
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import mean_absolute_error as mae
pd.options.display.max_rows = 50
## Install xlrd package to load Excel files
# conda install openpyxl
## conda install xlrd

## Neural Network 

## To  predict a new Amino Acid sequence/or its class (variant).

### Dataset

In [ ]:
test = pd.read_csv('new_for_model_testing.csv')
test= list(test['spike_aa'])[4]#for predicting 
#test#

modelling_set = pd.read_csv('modelling_set1.csv')#for trainingset

sequence=list(modelling_set['spike_aa'])#feature column
label=list(modelling_set['variant'])#variants as labels

#modelling_set = modelling_set.drop_duplicates(subset=['spike_aa']).reset_index()
#modelling_set

In [ ]:
#test.info()
test=pd.DataFrame(test[['spike_aa','variant']][3:17]).reset_index()#variant
test
test.to_csv('test_unseen_data.csv', index = False)

In [ ]:
modelling_set.columns

In [ ]:
modelling_set=modelling_set[['orf1a_aa', 'variant']]
modelling_set
modelling_set.to_csv("modelling_set.csv", index=False)

In [ ]:
#test=pd.DataFrame(test)

test.to_csv("test_sequence.csv", index=False)


In [ ]:
modelling_set=modelling_set[['orf1a_aa', 'variant']]
modelling_set

## Define: 
**1. Amino acids sequences**

**2. Characters present in the strings 'sequence'** and

**3. Function for the one-hot encode array**

In [ ]:
#chatGTP on 18:00 11.05.2023. It kills the python kernel. Too expensive for my computer
import numpy as np

# Define a list of amino acids in alphabetical order
amino_acids = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', '*']

# Define a function to convert a string of amino acids to a one-hot encoded array
def one_hot_encode(sequence, amino_acids):
    # Initialize an array of zeros with dimensions (length of sequence, number of amino acids)
    encoded = np.zeros((len(sequence), len(amino_acids)))
    # Loop over each amino acid in the sequence
    for i, amino_acid in enumerate(sequence):
        # Set the corresponding position in the encoded array to 1
        encoded[i, amino_acids.index(amino_acid)] = 1
    return encoded




## Define: 
**1. Training set of amino acid sequences** and 

**2. The corresponding labels (variants)**

In [ ]:
# Define a training set of amino acid sequences and their corresponding labels
#training_sequences = ['ACDEFGHIKLMNPQRSTVWY', 'AAAAAAACCCCCGGGGGTTTTT', 'ACDEFGHIKLMNPQRSTVWY', 'ACDEFGHIKLMNPQRSTVWY']
#training_labels = ['protein', 'nucleotide', 'protein', 'protein']
training_sequences = sequence#df['spike_aa']
training_labels = label#df['variant']


## Encode the training sequences as one-hot encoded arrays

In [ ]:
#Encode the training sequences as one-hot encoded arrays
encoded_training_sequences = np.array([one_hot_encode(sequence, amino_acids) for sequence in training_sequences])

# Define a simple neural network with one hidden layer
from keras.models import Sequential
from keras.layers import Dense

#model = Sequential()#binary class
model = tf.keras.models.Sequential()#For multi-class
model.add(Dense(units=10, input_dim=len(amino_acids), activation='relu'))
model.add(Dense(units=1, activation='sigmoid'))


## Compile the model with multi-class cross-entropy loss and Adam optimizer

In [ ]:
#model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])#binary

# Compile the model for multi-class cross-entropy loss function used for classification problems.
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])


## Train the model on the encoded training sequences and labels

In [ ]:
# Train the model on the encoded training sequences and labels
model.fit(encoded_training_sequences, np.array(training_labels)=='protein', epochs=100, batch_size=1)

# Define a test sequence and encode it as a one-hot encoded array
test_sequence = test['spike_aa']#'ACDEFGHIKLMNPQRSTVWY'
encoded_test_sequence = one_hot_encode(test_sequence, amino_acids)



## Predict the label of the test sequence using the trained model

In [ ]:
# Predict the label of the test sequence using the trained model
prediction = model.predict(np.array([encoded_test_sequence]))

# Print the predicted label
if prediction > 0.5:
    print('This is a known variant.')
else:
    print('This is likely to be a new variant.')

# Random Forest Classifier

## To  predict a new Amino Acid sequence/or its class (variant).

In [4]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')
## multiple outputs
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import minmax_scale
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder# example of a one hot encoding
from numpy import asarray# example of a one hot encoding
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PowerTransformer
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm
from scipy.stats import boxcox
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import mean_absolute_error as mae
pd.options.display.max_rows = 50
## Install xlrd package to load Excel files
# conda install openpyxl
## conda install xlrd

## Dataset

In [11]:
test = pd.read_csv('new_for_model_testing.csv')


test= list(test['spike_aa'])[4]#for predicting 
#test#
#test= test['spike_aa']
#modelling_set = pd.read_csv('modelling_set1.csv')#for trainingset
modelling_set = pd.read_csv('combined3melted_orf1a_s.csv')

#Using orf1a
#sequence=list(modelling_set['orf1a_aa'])#feature column
#label=list(modelling_set['variant'])#variants as labels

#modelling_set = modelling_set.drop_duplicates(subset=['spike_aa']).reset_index()
#modelling_set

In [12]:
modelling_set.columns

Index(['orf1a_cdna_width', 'orf1a_aa', 'orf1a_aa_width', 'spike_cdna_width',
       'spike_aa', 'spike_aa_width', 'B.1.1.7_Alpha', 'B.1.351_Beta',
       'B.1.1.248_P1_Gamma', 'P.2_Zeta', 'B.1.617.1_Kappa', 'B.1.617.2_Delta',
       'B.1.617.2_Delta_Plus', 'B.1.617.3', 'B.1.427_Epsilon',
       'B.1.429_Epsilon', 'B.1.526_Iota', 'B.1.525_Eta', 'P.3_Theta',
       'B.1.1.529_Omicron', 'B.1.1.529.1_Omicron', 'Delta_lineage',
       'Omicron_lineage', 'Other', 'variant', 'duplicated'],
      dtype='object')

In [13]:
modelling_set =modelling_set.drop_duplicates(subset=['spike_aa']).reset_index()
modelling_set =modelling_set[['spike_aa','spike_aa_width', 'variant']]
modelling_set

spike_aa  spike_aa_width  \
0      MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
1      MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
2      MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
3      MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
4      MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
...                                                  ...             ...   
12601  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
12602  MFVFLVLLPLVSSQCVNLITRTQSYTNSFTRGVYYPDKVFRSSVLH...            1267   
12603  MFVFLVLLPLVSIQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
12604  MFVFLVLLPLVFSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   
12605  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...            1270   

                 variant  
0          B.1.1.7_Alpha  
1                  Other  
2          B.1.1.7_Alpha  
3          B.1.1.7_Alpha  
4          B.1.1.7_Alpha  
...                  ...  
12601  B.1.1.529_Omicron  
12602              Other  
12603  B.1.1.529_Omicron  
12604  B.1.1.529_Omicron  
12605  B.1.1.529_Omicron  

[12606 rows x 3 columns]

In [14]:
#modelling_set.drop_duplicates()

## Define: 
**1. Amino acids sequences**

**2. Characters present in the strings 'sequence'** and

**3. Function for the one-hot encode array**

In [18]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
#
#Using spike
sequence=list(modelling_set['spike_aa'])#feature column
label=list(modelling_set['variant'])#variants as labels
#
#define amino acids possibilities found in the sequence
amino_acids = 'ACDEFGHIKLMNPQRSTVWY*'

# Define the amino acid sequence data
X=np.array(sequence)

# Define the corresponding labels (e.g., 0 = non-functional, 1 = functional)
y= label

max_sequence_length=modelling_set['spike_aa_width'].max()
num_samples =len(modelling_set)
# Convert amino acid sequences to numerical vectors
aa_to_index = {aa: i for i, aa in enumerate(amino_acids)}
X_encoded = []
for i in range(num_samples):
    encoded_seq = [aa_to_index[aa] for aa in X[i]]
    padded_seq = np.pad(encoded_seq, (0, max_sequence_length - len(encoded_seq)))
    X_encoded.append(padded_seq)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

# Train the Random Forest Classifier
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

# Evaluate the performance of the model
y_pred = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred)
print('Accuracy:', acc*100,"%")


RandomForestClassifier(random_state=42)

Accuracy: 99.52418715305313 %


In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

# Define the amino acid sequence data
sequences = sequence#np.array(sequence)#sequence


# Define the corresponding labels (e.g., 0 = non-functional, 1 = functional)
labels = label

# Define the 20 amino acids and the stop codon *
#amino_acids = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', '*']
amino_acids = 'ACDEFGHIKLMNPQRSTVWY*'
# Convert the amino acid sequences to numerical features using one-hot encoding
#features = []
#for seq in sequences:
 #   feature = []
  #  for aa in seq:
   #     if aa not in amino_acids:
    #        raise ValueError('Invalid amino acid character: %s' % aa)
     #   aa_idx = amino_acids.index(aa)
      #  aa_feature = [int(i == aa_idx) for i in range(len(amino_acids))]
       # feature.extend(aa_feature)
   # features.append(feature)
    #features.append(np.array(feature))
    
aa_to_index = {aa: i for i, aa in enumerate(amino_acids)}
X_encoded = []
for i in range(num_samples):
    encoded_seq = [aa_to_index[aa] for aa in X[i]]
    padded_seq = np.pad(encoded_seq, (0, max_sequence_length - len(encoded_seq)))
    X_encoded.append(padded_seq)
    
    

## Split the data into training and testing sets

In [20]:
# Split the data into training and testing sets
labels = label
#X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(X_encoded, labels, test_size=0.2, random_state=42)


#X_train = np.array(X_train).reshape(-1,1)

## Encode the training sequences as one-hot encoded arrays

In [21]:
# Train a random forest classifier on the training data
rfc = RandomForestClassifier(n_estimators=100, random_state=42)
rfc.fit(X_train, y_train)



RandomForestClassifier(random_state=42)

## Predict the labels of the testing data using the trained classifier

In [ ]:
# Predict the labels of the testing data using the trained classifier
y_pred = rfc.predict(X_test)



## Evaluate the accuracy of the classifier on the testing data

In [22]:
from sklearn.metrics import precision_score, recall_score, classification_report

In [23]:
np.array(X_train)[0]

array([10,  4, 17, ...,  0,  0,  0])

In [24]:
print(classification_report( y_test, y_pred))

                      precision    recall  f1-score   support

  B.1.1.248_P1_Gamma       0.80      1.00      0.89         8
   B.1.1.529_Omicron       0.99      0.99      0.99        78
       B.1.1.7_Alpha       1.00      1.00      1.00       412
     B.1.427_Epsilon       1.00      1.00      1.00        32
     B.1.617.2_Delta       1.00      1.00      1.00      1232
B.1.617.2_Delta_Plus       1.00      1.00      1.00         1
       Delta_lineage       1.00      0.92      0.96        50
     Omicron_lineage       1.00      0.96      0.98        24
               Other       0.99      1.00      0.99       677
            P.2_Zeta       0.83      0.62      0.71         8

            accuracy                           1.00      2522
           macro avg       0.96      0.95      0.95      2522
        weighted avg       1.00      1.00      1.00      2522



In [25]:
classification_report( y_test, y_pred)

'                      precision    recall  f1-score   support\n\n  B.1.1.248_P1_Gamma       0.80      1.00      0.89         8\n   B.1.1.529_Omicron       0.99      0.99      0.99        78\n       B.1.1.7_Alpha       1.00      1.00      1.00       412\n     B.1.427_Epsilon       1.00      1.00      1.00        32\n     B.1.617.2_Delta       1.00      1.00      1.00      1232\nB.1.617.2_Delta_Plus       1.00      1.00      1.00         1\n       Delta_lineage       1.00      0.92      0.96        50\n     Omicron_lineage       1.00      0.96      0.98        24\n               Other       0.99      1.00      0.99       677\n            P.2_Zeta       0.83      0.62      0.71         8\n\n            accuracy                           1.00      2522\n           macro avg       0.96      0.95      0.95      2522\n        weighted avg       1.00      1.00      1.00      2522\n'

In [28]:
# Evaluate the accuracy of the classifier on the testing data
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy*100,'%')

Accuracy: 99.52418715305313 %


In [29]:
# Evaluate the accuracy of the classifier on the testing data
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy*100,'%')



Accuracy: 99.52418715305313 %


In [30]:
# Evaluate the accuracy of the classifier on the testing data
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy*100,'%')



Accuracy: 99.52418715305313 %


In [31]:
# Evaluate the accuracy of the classifier on the testing data
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy*100,'%')


Accuracy: 99.52418715305313 %


In [32]:
# Evaluate the accuracy of the classifier on the testing data
accuracy = recall_score(y_test, y_pred, average="weighted")
print("Accuracy:", accuracy*100,'%')



Accuracy: 99.52418715305313 %


In [33]:
# Evaluate the accuracy of the classifier on the testing data
accuracy = precision_score(y_test, y_pred, average="weighted")
print("Accuracy:", accuracy*100,'%')



Accuracy: 99.52744944998335 %


# Predicting unseen Amino Acid Sequence

In [40]:
#test = pd.read_csv('new_for_model_testing2.csv')
test = pd.read_csv('new_for_model_testing.csv')
new_sequence= test['spike_aa'][4]
#new_sequence= list(test['spike_aa'])[700]

max_sequence_length=modelling_set['spike_aa_width'].max()
# Convert the new sequence to a numerical vector using the same encoding scheme as the training data
encoded_seq = [aa_to_index[aa] for aa in new_sequence]
padded_seq = np.pad(encoded_seq, (0, max_sequence_length - len(encoded_seq)))
X_new = [padded_seq]

# Use the trained classifier to predict the class of the new sequence
y_new = clf.predict(X_new)
print('Predicted class:', y_new[0])


Predicted class: B.1.1.7_Alpha


In [ ]:
sequence

# Random Forest to Predict the probability of having new mutations in a new protein sequence

In [56]:

from sklearn.ensemble import RandomForestClassifier
import numpy as np


# Define the training data
X_train = np.array(sequence)
y_train = y = label

# Define the encoding scheme
aa_list = 'ACDEFGHIKLMNPQRSTVWY*'
aa_to_index = {aa: i for i, aa in enumerate(aa_list)}

# Encode the training data
max_sequence_length = max([len(seq) for seq in X_train])
X_encoded = [[aa_to_index[aa] for aa in seq] for seq in X_train]
X_padded = [np.pad(seq, (0, max_sequence_length - len(seq))) for seq in X_encoded]

# Train the classifier
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_padded, y_train)

# Define a new sequence with a mutation and encode it
new_sequence = list(test['spike_aa'])[4]#'MENLC*MNVVRKF*KC'
new_encoded = [aa_to_index[aa] for aa in new_sequence]
new_padded = np.pad(new_encoded, (0, max_sequence_length - len(new_encoded)))

# Use the classifier to predict the probability of having new mutations
y_pred_proba = clf.predict_proba([new_padded])
print('Probability of each mutation class:', y_pred_proba[0]*100,'%')


Probability of each mutation class: [  0.   0. 100.   0.   0.   0.   0.   0.   0.   0.   0.] %


# Summary: Random Forest

In [ ]:
test = pd.read_csv('new_for_model_testing.csv')


#test= list(test['spike_aa'])[4]#for predicting 
#test#
#test= test['spike_aa']
#modelling_set = pd.read_csv('modelling_set1.csv')#for trainingset
modelling_set = pd.read_csv('trainingset.csv')
#Using spike
sequence=list(modelling_set['spike_aa'])#feature column
label=list(modelling_set['variant'])#variants as labels

#
#Using orf1a
#sequence=list(modelling_set['orf1a_aa'])#feature column
#label=list(modelling_set['variant'])#variants as labels

#modelling_set = modelling_set.drop_duplicates(subset=['spike_aa']).reset_index()
#modelling_set

In [ ]:
modelling_set

In [ ]:

#modelling_set =modelling_set.drop_duplicates(subset=['spike_aa'])
#modelling_set

In [ ]:
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

#Amino acid
amino_acids = 'ACDEFGHIKLMNPQRSTVWY*'

X = np.array(sequence)
y = label

# Convert amino acid sequences to numerical vectors
aa_to_index = {aa: i for i, aa in enumerate(amino_acids)}
X_encoded = []
for i in range(len(sequence)):
    encoded_seq = [aa_to_index[aa] for aa in X[i]]
    padded_seq = np.pad(encoded_seq, (0, modelling_set['spike_aa_width'].max() - len(encoded_seq)))
    X_encoded.append(padded_seq)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

# Train the Random Forest Classifier
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

# Evaluate the performance of the model
y_pred = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred)
print('Accuracy:', acc*100,'%')

In [ ]:
test

In [ ]:
#test = pd.read_csv('new_for_model_testing2.csv')
test = pd.read_csv('new_for_model_testing.csv')

#new_sequence= test['spike_aa']#[2]
new_sequence= list(test['spike_aa'])[13]


# Convert the new sequence to a numerical vector using the same encoding scheme as the training data
encoded_seq = [aa_to_index[aa] for aa in new_sequence]
padded_seq = np.pad(encoded_seq, (0, modelling_set['spike_aa_width'].max() - len(encoded_seq)))
X_new = [padded_seq]

# Use the trained classifier to predict the class of the new sequence
y_new = clf.predict(X_new)
print('Predicted class:', y_new[0])

In [ ]:
###With cDNA in spike gene

In [ ]:
test = pd.read_csv('new_for_model_testing.csv')


test= list(test['spike_cdna'])[4]#for predicting 
#test#
#test= test['spike_aa']
#modelling_set = pd.read_csv('modelling_set1.csv')#for trainingset
modelling_set = pd.read_csv('trainingset.csv')
#Using spike
sequence=list(modelling_set['spike_cdna'])#feature column
label=list(modelling_set['variant'])#variants as labels

#spike_cdna
#Using orf1a
#sequence=list(modelling_set['orf1a_aa'])#feature column
#label=list(modelling_set['variant'])#variants as labels

#modelling_set = modelling_set.drop_duplicates(subset=['spike_aa']).reset_index()
#modelling_set

In [ ]:
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

#Amino acid
cdna_seqs = 'ATCG'

X = np.array(sequence)
y = label

# Convert amino acid sequences to numerical vectors
cdna_to_index = {cdna: i for i, cdna in enumerate(cdna_seqs)}
X_encoded = []
for i in range(len(sequence)):
    encoded_seq = [cdna_to_index[cdna] for cdna in X[i]]
    padded_seq = np.pad(encoded_seq, (0, modelling_set['spike_cdna_width'].max() - len(encoded_seq)))
    X_encoded.append(padded_seq)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

# Train the Random Forest Classifier
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

# Evaluate the performance of the model
y_pred = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred)
print('Accuracy:', acc*100,'%')

In [ ]:
test

In [ ]:
test = pd.read_csv('new_for_model_testing.csv')
test

In [ ]:

new_sequence= test['spike_aa']#[2]
new_sequence= list(test['spike_aa'])[31]


# Convert the new sequence to a numerical vector using the same encoding scheme as the training data
encoded_seq = [aa_to_index[aa] for aa in new_sequence]
padded_seq = np.pad(encoded_seq, (0, modelling_set['spike_cdna_width'].max() - len(encoded_seq)))
X_new = [padded_seq]

# Use the trained classifier to predict the class of the new sequence
y_new = clf.predict(X_new)
print('Predicted class:', y_new[0])



Probability of each mutation class: [  0.   0. 100.   0.   0.   0.   0.   0.   0.   0.   0.] %


# Support Vector Machines (SVMs)

In [12]:
# Import necessary libraries
import pandas as pd
from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import CountVectorizer

# Load the dataset
data = pd.read_csv('trainingset.csv')

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data['spike_aa'], data['variant'], test_size=0.2, random_state=42)

# Preprocess the data
# You may need to implement custom preprocessing for your specific problem.

# Vectorize the amino acid sequences
vectorizer = CountVectorizer(analyzer='char', ngram_range=(1, 3))
X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

# Define the SVM model
clf = svm.SVC(kernel='linear', C=1, probability=True)

# Train the model
clf.fit(X_train, y_train)

# Predict the test set labels
y_pred = clf.predict(X_test)

# Calculate accuracy score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy*100,'%')


Accuracy: 99.66507177033492 %


In [31]:
# Predict the label of a new sequence
new_seq = list(test['spike_aa'])[8]#'MAAFVKVLVTLYLAVAVFVFNAKGEHR'
new_seq_vec = vectorizer.transform([new_seq])
predicted_label = clf.predict(new_seq_vec)[0]
print('Predicted label:', predicted_label)

Predicted label: Other


# Naive Bayes model to classify amino acid sequences

In [10]:
# Import necessary libraries
import pandas as pd
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import CountVectorizer

# Load the dataset
data = pd.read_csv('trainingset.csv')
data =data.drop_duplicates(subset=['spike_aa']).reset_index()
data =data[['spike_aa','spike_aa_width', 'variant']]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data['spike_aa'], data['variant'], test_size=0.2, random_state=42)

# Preprocess the data
# You may need to implement custom preprocessing for your specific problem.

# Vectorize the amino acid sequences
vectorizer = CountVectorizer(analyzer='char', ngram_range=(1,1))
X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

# Define the Naive Bayes model
clf = MultinomialNB()

# Train the model
clf.fit(X_train, y_train)

# Predict the test set labels
y_pred = clf.predict(X_test)

# Calculate accuracy score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy*100,'%')


Accuracy: 51.55795596177815 %


# Convolutional Neural Network (CNN)

In [4]:
test = pd.read_csv('new_for_model_testing.csv')


#test= list(test['spike_aa'])[4]#for predicting 
#test#
#test= test['spike_aa']
#modelling_set = pd.read_csv('modelling_set1.csv')#for trainingset
modelling_set = pd.read_csv('trainingset.csv')
modelling_set =modelling_set.drop_duplicates(subset=['spike_aa']).reset_index()
modelling_set =modelling_set[['spike_aa','spike_aa_width', 'variant']]
#Using spike
sequence=list(modelling_set['spike_aa'])#feature column
label=list(modelling_set['variant'])#variants as labels

#
#Using orf1a
#sequence=list(modelling_set['orf1a_aa'])#feature column
#label=list(modelling_set['variant'])#variants as labels

#modelling_set = modelling_set.drop_duplicates(subset=['spike_aa']).reset_index()
#modelling_set

In [ ]:
#Amino acid
amino_acids = 'ACDEFGHIKLMNPQRSTVWY*'

X = np.array(sequence)
y = label

# Convert amino acid sequences to numerical vectors
aa_to_index = {aa: i for i, aa in enumerate(amino_acids)}
X_encoded = []
for i in range(len(sequence)):
    encoded_seq = [aa_to_index[aa] for aa in X[i]]
    padded_seq = np.pad(encoded_seq, (0, modelling_set['spike_aa_width'].max() - len(encoded_seq)))
    X_encoded.append(padded_seq)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

# Train the Random Forest Classifier
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

# Evaluate the performance of the model
y_pred = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred)
print('Accuracy:', acc*100,'%')

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv1D, MaxPooling1D, Flatten, Dense, Dropout

# About the data
num_samples = len(modelling_set)
sequence_length = modelling_set['spike_aa_width'].max()
num_classes = 11
amino_acids = 'ACDEFGHIKLMNPQRSTVWY*'

#data. Amino acid and its labels(variant)
X = np.array(sequence)
y = label

# Convert amino acid sequences to one-hot encoding
aa_to_index = {aa: i for i, aa in enumerate(amino_acids)}
X_one_hot = np.zeros((num_samples, sequence_length, len(amino_acids)))
for i in range(num_samples):
    for j in range(sequence_length):
        aa_index = aa_to_index[X[i][j]]
        X_one_hot[i][j][aa_index] = 1

# Define the model architecture
model = Sequential()
model.add(Conv1D(filters=64, kernel_size=3, activation='relu', input_shape=(sequence_length, len(amino_acids))))
model.add(MaxPooling1D(pool_size=2))
model.add(Conv1D(filters=64, kernel_size=3, activation='relu'))
model.add(MaxPooling1D(pool_size=2))
model.add(Flatten())
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

# Compile the model
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
model.fit(X_one_hot, y, epochs=10, batch_size=32, validation_split=0.2)


# Recurrent Neural Network (RNN)

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense

# About the data
num_samples = len(modelling_set)
sequence_length = modelling_set['spike_aa_width'].max()
num_classes = 11
amino_acids = 'ACDEFGHIKLMNPQRSTVWY*'

#data. Amino acid and its labels(variant)
X = np.array(sequence)
y = label

# Convert amino acid sequences to one-hot encoding
aa_to_index = {aa: i for i, aa in enumerate(amino_acids)}
X_one_hot = np.zeros((num_samples, sequence_length, len(amino_acids)))
for i in range(num_samples):
    for j in range(sequence_length):
        aa_index = aa_to_index[X[i][j]]
        X_one_hot[i][j][aa_index] = 1

# Define the model architecture
model = Sequential()
model.add(LSTM(64, input_shape=(sequence_length, len(amino_acids))))
model.add(Dense(num_classes, activation='softmax'))

# Compile the model
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
model.fit(X_one_hot, y, epochs=10, batch_size=32, validation_split=0.2)


# Conclusion with different classifiers

### 1. Random Forest Classifier
1. Both the cDNA and Amino acid prediction with spike give >99.5% accuracy
2. But the prediction is correctly predicted with Amino acid seqs.
3. cDNA seq predctions give all the variants predicted on unseen sequence as 'other'.
4. So the better way is to use amino acid sequences for prediction of the variants and probability of getting having new mutations in a sequences consistent to each class (variant).
5. Overally, Random Forest gives the best prediction of >99.5%, and predicts correctly with spike amino acids, and can get the probability of getting new mutations.
6. Because of large sizes of sequences as string, reducing thenumber by dropping duplicates was helpful to reduce the size of file and time taken to computer and reduces the chances of running out of the memory of the computer.

### 2. Support Vector Machines (SVMs)
1. Gave an accuracy of 99.67%, and give the correct prediction on the correct sequences (new) 
2. But if the sequence of is wrong or dammy sequence, then it predicts it as omicron, which is wrong. 
3. Hence not robust in prediction of new sequences.

### 3. Naive Bayes
1. Accuracy of 51.6%
2. Hence not robust enough.

### 4. Neutral Network models
1. All the different neutral network models did not compute in my computer the python kernel died
2. Possible because of one-hot encoding that needed alot of comptations and increase in dimensionality of that data
3. Hence the memory of the computer became limited.
